src/main/host/memory_manager.rs

```diff
@@ -0,0 +1,769 @@
+        let heap = get_heap(&shm_file, thread, &mut regions);
+        let stack_end = map_stack(&shm_file, &mut regions);
+
+        MemoryManager {
+            shm_file: RefCell::new(shm_file),
+            regions: RefCell::new(regions),
+            misses_by_path: RefCell::new(HashMap::new()),
+            need_post_exec_cleanup: Cell::new(false),
+            heap: RefCell::new(heap),
+            stack_copied: RefCell::new(stack_end..stack_end),
+        }
+    }
+
+    // Clears all mappings, unmapping them from shadow's address space and recovering space from
+    // the memory file. e.g. called after execve and in Drop.
+    fn reset_and_unmap_all(&self) {
+        // Mappings are no longer valid. Clear out our map, and unmap those regions.
+        let mutations = self
+            .regions
+            .borrow_mut()
+            .clear(std::usize::MIN..std::usize::MAX);
+        for m in mutations {
+            match m {
+                Mutation::Removed(interval, region) => {
+                    if region.shadow_base != std::ptr::null_mut() {
+                        let res = unsafe {
+                            libc::munmap(region.shadow_base, interval.end - interval.start)
+                        };
+                        if res != 0 {
+                            println!("Warning: munmap failed");
+                        }
+                    }
+                }
+                _ => (),
+            }
+        }
+        self.shm_file.borrow_mut().truncate();
+    }
+
+    /// Should be called by shadow's SysCallHandler before allowing the plugin to execute the exec
+    /// syscall.
+    pub fn pre_exec_hook(&mut self, _thread: &impl Thread) {
+        self.need_post_exec_cleanup.set(true);
+    }
+
+    // Called internally on the next usage *after* execve syscall has executed. Re-initializes as
+    // needed.
+    fn post_exec_cleanup_if_needed(&self, thread: &impl Thread) {
+        if !self.need_post_exec_cleanup.get() {
+            return;
+        }
+        self.reset_and_unmap_all();
+        let shm_file = self.shm_file.borrow();
+        *self.regions.borrow_mut() = get_regions(thread.get_system_pid());
+        *self.heap.borrow_mut() = get_heap(&*shm_file, thread, &mut *self.regions.borrow_mut());
+        let stack_end = map_stack(&*shm_file, &mut *self.regions.borrow_mut());
+        *self.stack_copied.borrow_mut() = stack_end..stack_end;
+        self.need_post_exec_cleanup.set(false);
+    }
+
+    /// Gets a reference to an object in the plugin's memory.
+    /// SAFETY
+    /// * The pointer must point to readable value of type T.
+    /// * Returned ref mustn't be accessed after Thread runs again or flush is called.
+    #[allow(dead_code)]
+    pub unsafe fn _get_ref<T>(&self, thread: &impl Thread, src: PluginPtr) -> Result<&T, i32> {
+        let raw = self.get_readable_ptr(thread, src, std::mem::size_of::<T>())?;
+        Ok(&*(raw as *const T))
```

Comment:
I think there should be a check here that raw is not null, since as far as I can tell `get_readable_ptr()` can return a null pointer, which shouldn't be cast as a rust reference. (And the same for functions below, including the `from_raw_parts()`.)